lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+/**
+ * @license LGPLv3, http://opensource.org/licenses/LGPL-3.0
+ * @copyright Aimeos (aimeos.org), 2015-2020
+ * @package MShop
+ * @subpackage Service
+ */
+
+
+namespace Aimeos\MShop\Service\Provider\Payment;
+
+
+use Omnipay\Omnipay as OPay;
+use Aimeos\MShop\Order\Item\Base as Status;
+
+use Illuminate\Support\Facades\Auth;
+
+
+/**
+ * Payment provider for payment gateways supported by the PaypalPlus library.
+ *
+ * @package MShop
+ * @subpackage Service
+ */
+class PaypalPlus
+extends    \Aimeos\MShop\Service\Provider\Payment\OmniPay
+//	extends \Aimeos\MShop\Service\Provider\Payment\Base
+	implements \Aimeos\MShop\Service\Provider\Payment\Iface
+{
+
+	protected function processOrder( \Aimeos\MShop\Order\Item\Iface $order,
+		array $params = [] ) : ?\Aimeos\MShop\Common\Helper\Form\Iface
+	{
+		$parts = \Aimeos\MShop\Order\Item\Base\Base::PARTS_SERVICE
+			| \Aimeos\MShop\Order\Item\Base\Base::PARTS_PRODUCT
+			| \Aimeos\MShop\Order\Item\Base\Base::PARTS_ADDRESS;
+
+		$base = $this->getOrderBase( $order->getBaseId(), $parts );
+		$data = $this->getData( $base, $order->getId(), $params );
+		$urls = $this->getPaymentUrls();
+
+		$data['cancelUrl'] = $this->getConfigValue( 'cancelUrl', '' ) ;
+
+		if(empty($data['cancelUrl'])) {
+			$data['cancelUrl'] = $_SERVER['HTTP_ORIGIN'] ;
+		}
```

Comment:
The cancelUrl is incorrect. If it's required by paypal, it's the same as 'payment.url-success'. There, the customer can choose to retry or change the payment

Reply:
but it keeps cancelUrl as returnUrl unless I give it value as in my code !

Reply:
Usually, that's desired. The success and cancel URLs both point to checkout confirm page where the customer has the option to retry or change the payment (also to non-PayPal options)

Reply:
done , but how come that approved !
client will be forwarded to confirmation page and empty his basket ! 
any way .. I have commented out the code that set cancelUrl just to get sure from you about solving it